zebra-state/src/service/read/difficulty.rs

```diff
@@ -18,33 +19,85 @@ use crate::{
             AdjustedDifficulty,
         },
         finalized_state::ZebraDb,
+        read::tree::history_tree,
         NonFinalizedState,
     },
-    GetBlockTemplateChainInfo,
+    BoxError, GetBlockTemplateChainInfo, HashOrHeight,
 };
 
+use crate::service::block_iter::Iter;
+
 /// Returns the [`GetBlockTemplateChainInfo`] for the current best chain.
 ///
 /// # Panics
 ///
-/// If we don't have enough blocks in the state.
-pub fn difficulty_and_time_info(
+/// - If we don't have enough blocks in the state.
+/// - If a consistency check fails `RETRIES` times.
+pub fn get_block_template_chain_info(
     non_finalized_state: &NonFinalizedState,
     db: &ZebraDb,
     tip: (Height, Hash),
     network: Network,
 ) -> GetBlockTemplateChainInfo {
-    let relevant_chain = any_ancestor_blocks(non_finalized_state, db, tip.1);
-    difficulty_and_time(relevant_chain, tip, network)
+    // Amount of retries for a consistency check.
+    const RETRIES: usize = 3;
+
+    let mut relevant_chain_and_history_tree_result =
+        relevant_chain_and_history_tree(non_finalized_state, db, tip.1);
+
+    for _ in 0..RETRIES {
+        if relevant_chain_and_history_tree_result.is_ok() {
+            break;
+        }
+
+        relevant_chain_and_history_tree_result =
+            relevant_chain_and_history_tree(non_finalized_state, db, tip.1);
+    }
+    // remove the unwrap somehow
+    let (relevant_chain, history_tree) =
+        relevant_chain_and_history_tree_result.expect("consistency check failed");
+
+    difficulty_time_and_history_tree(relevant_chain, tip, network, history_tree)
```

Comment:
`difficulty_time_and_history_tree()` gets blocks from the state via `relevant_chain`, so it needs to be inside the tip check, or we need to explain why it doesn't need the tip check.

If we call `difficulty_time_and_history_tree()` inside the tip check, then all the request code will be inside it. So there's no possible way to access the state without checking the tip.

Reply:
Sure, but I'd like you to do the testing.

Reply:
Fixed in commit 2fb92f0 by doing the iterator collect inside the tip check. This also simplifies the types and lifetimes.